modules/docs/markdown/04-codegen/01-customisation/10-service-product.md

```diff
@@ -0,0 +1,217 @@
+
+def descriptor[Alg[_[_, _, _, _, _]]](mirror: ServiceProduct.Mirror[Alg]): mirror.Prod[Describe] =
+  mirror
+    .serviceProduct
+    .mapK5(
+      mirror.serviceProduct.endpointsProduct,
+      new PolyFunction5[mirror.serviceProduct.service.Endpoint, Describe] {
+
+        override def apply[I, E, O, SI, SO](
+          fa: mirror.serviceProduct.service.Endpoint[I, E, O, SI, SO]
+        ): Describe[I, E, O, SI, SO] =
+          s"def ${fa.name}(input: ${fa.input.shapeId.name}): ${fa.output.shapeId.name}"
+
+      },
+    )
+
+// Usage
+
+val desc: String = descriptor(ExampleService).exampleOperation
+```
+
+### Non-linear input of operation
+
+```scala mdoc
+import smithy4s.ShapeId
+
+type Id[A] = A
+
+val impl: ExampleService[Id] =
+  new ExampleService[Id] {
+
+    override def exampleOperation(input: String): ExampleOutput = ExampleOutput(
+      s"Output for $input!"
+    )
+
+  }
+
+type ListClient[I, _, O, _, _] = List[I] => List[O]
+
+def listClient[Alg[_[_, _, _, _, _]], Prod[_[_, _, _, _, _]]](
+  impl: smithy4s.kinds.FunctorAlgebra[Alg, Id]
+)(
+  implicit sp: ServiceProduct.Aux[Prod, Alg]
+): Prod[ListClient] = sp
+  .mapK5(
+    sp.endpointsProduct,
+    new PolyFunction5[sp.service.Endpoint, ListClient] {
+
+      private val interp = sp.service.toPolyFunction(impl)
+
+      override def apply[I, E, O, SI, SO](
+        fa: sp.service.Endpoint[I, E, O, SI, SO]
+      ): List[I] => List[O] = _.map(in => interp(fa.wrap(in)))
+
+    },
+  )
+
+// Usage
+
+listClient(impl)( /* implicit scope problem here - TODO */ ExampleService.serviceProduct)
+  .exampleOperation(
+    List("a", "b", "c").map(ExampleInput(_))
+  )
+  .mkString("\n")
+```
+
+### Fluent service builder
+
+```scala mdoc
+
+type ToList[_, _, O, _, _] = List[O]
+
+trait EndpointHandlerBuilder[I, E, O, SI, SO] {
+  def apply(f: I => O): EndpointHandler
+}
+
+sealed trait EndpointHandler {
+  type I_
+  type O_
+  def id: ShapeId
+  def function: I_ => O_
+}
+
+case class PartialBuilder[Alg[_[_, _, _, _, _]], Prod[_[_, _, _, _, _]]](
+  mirror: ServiceProduct.Mirror.Aux[Alg, Prod],
+  handlers: List[EndpointHandler],
+) {
+  private val sp: ServiceProduct.Aux[Prod, Alg] = mirror.serviceProduct
+
+  private val ehbProduct = sp
+    .mapK5(
+      sp.endpointsProduct,
+      new PolyFunction5[sp.service.Endpoint, EndpointHandlerBuilder] {
+
+        override def apply[I, E, O, SI, SO](
+          fa: sp.service.Endpoint[I, E, O, SI, SO]
+        ): EndpointHandlerBuilder[I, E, O, SI, SO] =
+          new EndpointHandlerBuilder[I, E, O, SI, SO] {
+
+            override def apply(f: I => O): EndpointHandler =
+              new EndpointHandler {
+                type I_ = I
+                type O_ = O
+                override val id: ShapeId = fa.id
+                override val function: I_ => O_ = f
+              }
+
+          }
+
+      },
+    )
+
+  def build: Alg[ToList] = sp
+    .service
+    .algebra(new sp.service.EndpointCompiler[ToList] {
```

Comment:
**praise:** for @Baccata - the whole `algebra` / `EndpointCompiler` thing is very cool.